File: h0rsescript-main/src/nativeMain/kotlin/h0rsescript/parser/Tokenizer.kt
package me.flaming.h0rsescript.parser

import me.flaming.h0rsescript.core.ErrorHandler
import me.flaming.h0rsescript.errors.InvalidTokenError

object Tokenizer {
    private const val KPREFIX = '$'
    private val keywords = listOf("define", "end", "parameters", "return", "mode", "include").map { k -> "$KPREFIX$k"}
    private val booleans = listOf("TRUE", "FALSE")
    private val symbols = listOf("->", "<->", ">", "<-", "<")

    var src = ""
    private var position = 0

    fun tokenize(input: String): MutableList<Token> {
        src = input
        val tokens = mutableListOf<Token>()

        val consumeToken = { token: Token ->
            tokens.add(token)
            position += token.value.length
        }

        while (position < input.length) {
            val char = input[position]
            val startPos = position

            // Handle identifiers, and booleans
            if (char.isLetter() || char == '_') {
                val value = getIdentifier()
                var tokenType = if (value.contains('.')) TokenType.QUALIFIED_IDENTIFIER else TokenType.IDENTIFIER

                // Handle booleans
                if (value in booleans) tokenType = TokenType.BOOLEAN

                tokens.add(Token(tokenType, value, position = startPos))
            }

            // Handle keywords
            else if (char == KPREFIX) {
                tokens.add(Token(TokenType.KEYWORD, getKeyword(), position = startPos))
            }

            // Handle assignment operators
            else if (isAssignmentOpChar()) {
                tokens.add(Token(TokenType.ASSIGNMENT_OPERATOR, getAssignmentOp(), position = startPos))
            }

            // Handle strings and numbers
            else if (char == '"') {
                tokens.add(Token(TokenType.STRING, getStringLiteral(), position = startPos))
            }
            else if (char == '-' || char.isDigit()) {
                tokens.add(Token(TokenType.NUMBER, getNumberLiteral(), position = startPos))
            }

            // Handle single character tokens like brackets and commas and whitespaces
            else if (char == '[') {
                consumeToken(Token(TokenType.OPEN_BRACKET, "[", position = startPos))
            } else if (char == ']') {
                consumeToken(Token(TokenType.CLOSE_BRACKET, "]", position = startPos))
            } else if (char == '{') {
                consumeToken(Token(TokenType.OPEN_CURLY, "{", position = startPos))
            } else if (char == '}') {
                consumeToken(Token(TokenType.CLOSE_CURLY, "}", position = startPos))
            } else if (char == ',') {
                consumeToken(Token(TokenType.COMMA, ",", position = startPos))
            } else if (char.isWhitespace()) {
                position++
            }

            // Comments
            else if (char == '#') {
                tokens.add(Token(TokenType.COMMENT, getComment(), position = startPos))
            }

            // Throw InvalidTokenError
            else {
                ErrorHandler.report(InvalidTokenError(src[position], getLineCol(position).first, getLineCol(position).second))
            }
        }
        return tokens
    }

    // Goofy ahh code
    private fun isAssignmentOpChar(): Boolean {
        val subStr = src.substring(position)
        return subStr.startsWith("<->") ||
                subStr.startsWith("->") ||
                subStr.startsWith("<-") ||
                subStr.startsWith(">") ||
                subStr.startsWith("<")
    }
    private fun getAssignmentOp(): String {
        val subStr = src.substring(position)
        var operatorValue = ""

        // Goofy ahh code v2
        if (subStr.startsWith("<->")) {
            operatorValue = "<->"
            position += 3
        } else if (subStr.startsWith("->")) {
            operatorValue += "->"
            position += 2
        } else if (subStr.startsWith("<-")) {
            operatorValue += "<-"
            position += 2
        } else if (subStr.startsWith(">")) {
            operatorValue += ">"
            position += 1
        } else if (subStr.startsWith("<")) {
            operatorValue += "<"
            position += 1
        } else {
            ErrorHandler.report(InvalidTokenError(src[position], getLineCol(position).first, getLineCol(position).second))
        }

        return operatorValue
    }

    private fun getStringLiteral(): String {
        val startPos = position
        var literalValue = ""

        // Start double quote "
        position++

        // Get all characters after opening ", until ending " or EOF
        while (!endOfSrc() && src[position] != '"') {
            literalValue += src[position]
            position++

            // Handle escape sequences
            if (src[position] == '\\') {
                literalValue += src[position++]
            }
        }

        // Ending double quote ", throw error if EOF
        if (endOfSrc()) {
            val err = InvalidTokenError(src[startPos], getLineCol(startPos).first, getLineCol(startPos).second)
            err.message += "\nUnterminated string literal"
            ErrorHandler.report(err)
        }
        position++

        return literalValue
    }

    // -?[0-9]+(\.[0-9]+)?
    private fun getNumberLiteral(): String {
        // Get first digit or negative symbol
        var literalValue = "${src[position]}"
        var isDecimalValid: Boolean? = null

        // Making decimal usage valid after 1 digit
        if (src[position].isDigit()) isDecimalValid = true

        position++

        // Get all numbers (and 1 decimal point)
        while (!endOfSrc() && (src[position].isDigit() || (src[position] == '.' && isDecimalValid == true))) {
            literalValue += src[position]

            // Making decimal usage valid after 1 digit
            if (isDecimalValid == null) isDecimalValid = true

            // Making decimal usage invalid after 1 use
            if (src[position] == '.') {
                isDecimalValid = false
            }

            position++
        }

        // Remove decimal if no number after it
        if (literalValue.endsWith('.')) {
            literalValue = literalValue.removeSuffix(".")
            position--
        }

        return literalValue
    }

    // [a-zA-Z_][a-zA-Z0-9_]*
    private fun getIdentifier(): String {
        // Get first letter or underscore
        var value = "${src[position]}"
        position++

        var endOfIdent = false
        while (!endOfIdent) {
            // Get current character, except .
            if (!endOfSrc() && isValidIdentChar(src[position])) {
                value += src[position]
                position++
            }
            // Get . only if it is followed by another character
            else if (!endOfSrc() && src[position] == '.' && (src.getOrNull(position + 1) != null && isValidIdentChar(src[position + 1]))) {
                value += src[position]
                position++
            }
            else endOfIdent = true
        }

        return value
    }

    private fun getKeyword(): String {
        // Skip first $
        val startPos = position
        var keyword = "${src[position]}"
        position++

        // Get only letters
        while (!endOfSrc() && src[position].isLetter()) {
            keyword += src[position]
            position++
        }

        // Throw error if keyword is invalid
        if (keyword !in keywords) {
            val err = InvalidTokenError(src[startPos], getLineCol(startPos).first, getLineCol(startPos).second)
            err.message += "\nInvalid keyword '$keyword' used"
            ErrorHandler.report(err)
        }

        return keyword
    }

    private fun getComment(): String {
        var comment = ""
        while (endOfSrc() || src[position] != '\n') {
            comment += src[position]
            position++
        }
        return comment
    }

    private fun isValidIdentChar(char: Char) = char.isLetterOrDigit() || char == '_'

    private fun endOfSrc() = position >= src.length

    fun getLineCol(pos: Int): Pair<Int, Int> {
        val subStr = src.substring(0, pos)
        val lines = subStr.split('\n')

        val line = lines.size
        val column = lines.last().length + 1

        return Pair(line, column)
    }
}

File: h0rsescript/src/parser/Tokenizer.cs
namespace h0rsescript.parser
{
    static class Tokenizer
    {
        // Initialize class
        private static string Src = "";
        private static int Pos = 0;
        private static List<Token> Tokens = [];
        
        // Helper properties
        private static char Current
        {
            get => Src[Pos];
        }
        private static char Previous
        {
            get => Src[Pos - 1];
        }
        private static bool IsEOF() => Pos >= Src.Length;
        private static char? Next
        {
            get => Pos < Src.Length ? Src[Pos + 1] : null;
        }

        // Language-defined constants
        private const char KeywordPrefix = '$';
        private static readonly string[] Keywords =
            (new string[] { "define", "end", "parameters", "return", "mode", "include" }).Select(k => KeywordPrefix + k).ToArray();
        private static readonly string[] AssignmentOperators =
            { "->", "<->", ">", "<-", "<" };
        private static readonly string[] Booleans =
            { "TRUE", "FALSE" };

        // Consume token
        private static void Consume(Func<char?> getTokenValue, TokenType tokenType, char? lastTokenToValidate = null)
        {
            int startPos = Pos;
            // Consume current character and move on
            string val = Current.ToString();
            Pos++;

            // Consume subsequent characters until null
            while (!IsEOF() && getTokenValue() != null)
            {
                val += getTokenValue();
                Pos++;
            }
            // Consume last character if it matches lastValidToken
            if (!IsEOF() && lastTokenToValidate != null && Current == lastTokenToValidate)
            {
                val += Current;
                Pos++;
            } else
            {
                // Throw Error
                // "Unexpected EOF at position {Pos-1}, where {lastValidToken} was expected"
            }

            Tokens.Add(new Token(Src, val, startPos, tokenType));
        }

        // Consume token and validate from a set of values
        private static void ConsumeAndValidate(Func<char?> getTokenValue, TokenType tokenType, string[] validValues)
        {
            Consume(getTokenValue, tokenType);
            
            // Validate token
            if (!validValues.Contains(Tokens.Last().Val))
            {
                // THROW ERROR
                Console.WriteLine($"WEEWOO1 {Tokens.Last().Val} at {Tokens.Last().Pos}");
                
            }
        }

        // Tokenize
        public static List<Token> Tokenize(string Str)
        {
            Src = Str;
            while (!IsEOF())
            {
                // Identifiers, Booleans and Qualified Identifiers
                if (char.IsLetter(Current) || Current == '_')
                {
                    Consume(GetIdentifier, TokenType.IDENTIFIER);
                    // Handle Booleans
                    if (Booleans.Contains(Tokens.Last().Val))
                    {
                        Token boolToken = Tokens.Last() with { Type = TokenType.BOOLEAN };
                        Tokens[^1] = boolToken;
                    }
                    // Handle Qualified Identifiers
                    if (Tokens.Last().Val.Contains('.'))
                    {
                        Token qIdentifierToken = Tokens.Last() with { Type = TokenType.QUALIFIED_IDENTIFIER };
                        Tokens[^1] = qIdentifierToken;
                    }
                }
                // String Literal
                else if (Current == '"')
                {
                    Consume(GetStringLiteral, TokenType.STRING, '"');
                }
                // Number Literal
                else if (char.IsDigit(Current) || Current == '-' && char.IsDigit(Next.GetValueOrDefault()))
                {
                    Consume(GetNumberLiteral, TokenType.NUMBER);
                }
                // Keywords
                else if (Current == KeywordPrefix)
                {
                    ConsumeAndValidate(GetIdentifier, TokenType.KEYWORD, Keywords);
                }
                // Assignment Operators
                else if (Current == '<' || Current == '-' || Current == '>')
                {
                    ConsumeAndValidate(GetAssignmentOperator, TokenType.ASSIGNMENT_OPERATOR, AssignmentOperators);
                }
                // Brackets & Commas
                else if (Current == '[')
                {
                    Consume(GetNull, TokenType.OPEN_BRACKET);
                }
                else if (Current == ']')
                {
                    Consume(GetNull, TokenType.CLOSE_BRACKET);
                }
                else if (Current == '{')
                {
                    Consume(GetNull, TokenType.OPEN_CURLY);
                }
                else if (Current == '}')
                {
                    Consume(GetNull, TokenType.CLOSE_CURLY);
                }
                else if (Current == ',')
                {
                    Consume(GetNull, TokenType.COMMA);
                }
                // Whitespace
                else if (char.IsWhiteSpace(Current))
                {
                    Pos++;
                }

                // Comments
                else if (Current == '#')
                {
                    Consume(GetComment, TokenType.COMMENT);
                }
                else
                {
                    // Throw InvalidTokenError TODO
                    Console.WriteLine($"WEEWOO `{Current}`");
                }
            }
            return Tokens;
        }

        // Token helper functions return the current char/null if a certain condition is true/false

        // Identifier : Returns all characters that matches the regex [a-zA-Z0-9_]
        private static char? GetIdentifier()
        {
            Func<char, bool> isValidIdentifierChar = (c) => char.IsLetterOrDigit(c) || c == '_';

            // GetValueOrDefault converts null to null char '\0' which is not a valid identifier char
            if (isValidIdentifierChar(Current) || (Current == '.' && isValidIdentifierChar(Next.GetValueOrDefault())))
                return Current;
            else
                return null;
        }

        // Assignment Operator : Returns the only possible subsequent characters i.e '-' and '>'
        // RIP "Goofy ahh code" 29 Dec 2024 - 14 Jun 2026
        private static char? GetAssignmentOperator()
        {
            if (Current == '-' || Current == '>')
                return Current;
            else
                return null;
        }

        // String : Returns all characters up to `"`
        private static char? GetStringLiteral()
        {
            // Handles escape sequences with `"`
            if (Current != '"' || (Current == '"' && Previous == '\\'))
                return Current;
            else
                return null;
        }

        // Number : Returns the digits 0-9 or '.' if a digit exists after it
        private static char? GetNumberLiteral()
        {
            // GetValueOrDefault converts null to null char '\0' which is not a digit
            if (char.IsDigit(Current) || (Current == '.' && char.IsDigit(Next.GetValueOrDefault())))
                return Current;
            else
                return null;
        }
        // Comment : Returns all characters except line break `\n`
        private static char? GetComment()
        {
            if (Current != '\n')
                return Current;
            else
                return null;
        }

        // GetNull : Used single char tokens that have no subsequent chars to consume
        private static char? GetNull() => null;
    }

    // Token & TokenType classes
    public record Token(string Src, string Val, int Pos, TokenType Type)
    {
        public override string ToString()
        {
            return $"Token {{ Val: {Val}, Pos: {Pos}, Type: {Type} }}";
        }
    };
    public enum TokenType
    {
        IDENTIFIER,    // Variable or function names
        QUALIFIED_IDENTIFIER, // Function signatures like math.add
        KEYWORD,       // Keywords like $define, $end, etc.

        ASSIGNMENT_OPERATOR,        // Symbols like ->, <->, etc.

        STRING,        // String literals
        NUMBER,        // Numeric literals
        BOOLEAN,       // TRUE/FALSE

        OPEN_BRACKET,  // [
        CLOSE_BRACKET, // ]
        OPEN_CURLY,    // {
        CLOSE_CURLY,   // }
        COMMA,         // ,

        COMMENT,       // Comments
    }

}
